lib/semantics/check-do-stmt.cc

```diff
@@ -344,41 +372,101 @@ static CS GatherReferencesFromExpression(const parser::Expr &expression) {
   }
 }
 
-// Find a canonical DO CONCURRENT and enforce semantics checks on its body
-class DoConcurrentContext {
+// Find a DO statement and enforce semantics checks on its body
+class DoStmtContext {
 public:
-  DoConcurrentContext(SemanticsContext &context)
-    : messages_{context.messages()} {}
-
-  bool operator==(const DoConcurrentContext &x) const { return this == &x; }
+  DoStmtContext(SemanticsContext &context) : context_{context} {}
 
   void Check(const parser::DoConstruct &doConstruct) {
+    if (doConstruct.IsDoConcurrent()) {
+      CheckDoConcurrent(doConstruct);
+      return;
+    }
+    if (doConstruct.IsDoNormal()) {
+      CheckDoNormal(doConstruct);
+      return;
+    }
+    // TODO: handle the other cases
+  }
+
+private:
+  using Bounds = parser::LoopControl::Bounds;
+
+  const Bounds &GetBounds(const parser::DoConstruct &doConstruct) {
+    auto &loopControl{doConstruct.GetLoopControl().value()};
+    return std::get<Bounds>(loopControl.u);
+  }
+
+  void CheckDoControl(parser::CharBlock sourceLocation, bool isReal) {
+    bool warn = context_.warnOnNonstandardUsage() ||
```

Comment:
```
  if (isReal && context_.ShouldWarn(parser::LanguageFeature::RealDoControls)) {
    context_.say(sourceLocation, "DO controls should be INTEGER"_en_US);
  }
```

That's it.  You don't have to check for a disabled feature (it's not one of those few) and you don't have to check whether warnings should be errors -- they will be converted to errors automatically if that's the case.

Reply:
Got it.  I'll make the changes.